packages/toolpad-app/src/toolpad/AppEditor/PageEditor/UrlQueryEditor.tsx

```diff
@@ -19,6 +20,38 @@ export interface UrlQueryEditorProps {
   pageNodeId: NodeId;
 }
 
+interface UrlQueryStringProps {
+  input: [string, string][] | undefined;
+}
+
+function UrlQueryString({ input }: UrlQueryStringProps) {
+  const [queryString, setQueryString] = React.useState('');
```

Comment:
* Always prefer deriving state from input. Setting state in an effect causes extra renders and doesn't initialize the correct initial value.
* Avoid constructing urls with string manipulation when you can.

```tsx
const queryString = React.useMemo(() => {
  const search = new URLSearchParams(input).toString();
  return search.size > 0 ? `?${search}` : '';
}, [input])
```